### PR TITLE
Add command: mlab_service_discovery for AppEngine Flex VM service discovery

### PR DESCRIPTION
**Pull request by stephen-soltesz**
This command adds an initial version of a special-purpose service discovery daemon that permits us to discover the public IP addresses of AppEngine Flex VMs for the purpose of scraping Prometheus metrics.

This initial version uses the environment's `DefaultCredentials` to query the AppEngine Admin API for a list of all VMs with public IPs and a forwarded port. This information is written to an output file in a format suitable for the file based service discovery mechanism. [https://prometheus.io/docs/operating/configuration/#<file_sd_config>](https://prometheus.io/docs/operating/configuration/#<file_sd_config>)

Future versions should run continuously and support additional sources, as well as a docker image configuration so that the daemon can be deployed easily as part of a k8s pod.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/prometheus-support/19)

### CODE REVIEW

File: cmd/mlab_service_discovery/main.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+	defaultScopes = []string{appengine.CloudPlatformScope, appengine.AppengineAdminScope}
+)
+
+var (
+	project   = flag.String("project", "mlab-sandbox", "project name")
+	filename  = flag.String("filename", "output.json", "File name to write output targets file.")
+	scopeList = flag.String("scopes", strings.Join(defaultScopes, ","), "Comma separated list of scopes to use.")
+)
+
+// discovery bundles several
```
Comment (gfr10598) [inline cmd/mlab_service_discovery/main.go]: Comment broken?

File: cmd/mlab_service_discovery/main.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+// formatTargets generates a JSON string from the slice of maps provided.
+func formatTargets(targets interface{}) string {
+	b, err := json.MarshalIndent(targets, "", "    ")
+	if err != nil {
+		fmt.Println("error:", err)
+	}
+	return string(b)
+}
+
+// newClient returns a new discovery struct with an authenticated AppEngine client.
```
Comment (gfr10598) [inline cmd/mlab_service_discovery/main.go]: instance or object instead of "struct"

File: cmd/mlab_service_discovery/main.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		"labels":  labels,
+		"targets": targets,
+	}
+	return values
+}
+
+// listVms walks through every AppEngine service, looks at every serving
+// version. listVms returns a list of every running instance in a form suitable
+// for export to a prometheus service discovery file.
+func listVms(client *discovery) ([]map[string]interface{}, error) {
```
Comment (gfr10598) [inline cmd/mlab_service_discovery/main.go]: Nesting is too deep.  Ok for quick start, but would prefer you extract functions in the next one or  two PRs.

File: cmd/mlab_service_discovery/main.go (hunk 1 above)
Comment (stephen-soltesz) [inline cmd/mlab_service_discovery/main.go]: Fixed.

File: cmd/mlab_service_discovery/main.go (hunk 2 above)
Comment (stephen-soltesz) [inline cmd/mlab_service_discovery/main.go]: Fixed.

File: cmd/mlab_service_discovery/main.go (hunk 3 above)
Comment (stephen-soltesz) [inline cmd/mlab_service_discovery/main.go]: I agree. I'll fix this next.

Comment (stephen-soltesz) [inline cmd/mlab_service_discovery/main.go]: Added a TODO.

**Pull request merged**